Fix the fix of ContestationPeriod fromDiffTime

ch1bo opened the pull request:
I think I need a break.. this time I added a test to make sure that we do not create `0` second contestation periods even with sub-second diff times.

This now also loads the parameters from the shelley genesis.json to not need to synchronize the network to get the `blockTime` and make this testable.

Also, this uses `NominalDiffTime` in more places (which is the better type for fixed durations).

---

<!-- Consider each and tick it off one way or the other -->
* [x] CHANGELOG updated or not needed
* [x] Documentation updated or not needed
* [x] Haddocks updated or not needed
* [x] No new TODOs introduced or explained herafter

ch1bo committed:
Fix the fix of ContestationPeriod fromDiffTime

abailly committed:
Simplify rounding to use ceiling

also cleanup some imports

ghost commented:
I took the liberty of directly pushing a change which I think is clearer, feel free to drop the commit

ch1bo commented:
Wait.... why is it `0.05s` in the first place.. It should be `20s`.

Need to investigate, there more is broken here.

ch1bo committed:
Define a test suite for withCardanoNodeOnKnownNetwork

This also asserts correctly computed block time now.

ch1bo committed:
Move waitForFullySynchronized and use RunningNode

The contained TODO can be addressed in a next step now.

ch1bo committed:
Read block time from shelley genesis JSON

This file must be available to start the cardano-node and we do not need
to wait for the network to be in a shelley compatible era to query it.

ch1bo committed:
Change withCardanoNode to read networkId from shelley genesis

This removes the hard-coded 42 network magic from withCardanoNodeDevnet

ch1bo committed:
Use NominalDiffTime in RunningNode blockTime

This is generally a more desirable type to describe a given duration of
time, e.g. how long to wait.

Updating our wait.. functions to NominalDiffTime also removed some
conversions in test code (realToFrac).

ch1bo commented:
> LGTM. Sorry for not paying enough attention but what was the rationale for moving from `DiffTime` to `NominalDiffTime` again?

`NominalDiffTime` is better suited for denoting simple durations, e.g. number of seconds to wait, and has functions working with it and `UTCTime` (the most common used type for denoting point in times). The only short-coming is that it's not complete at measuring time that includes leap seconds, e.g. for an exact number of seconds passed over multiple months. https://wiki.haskell.org/Time#Data_Types_for_Measuring_Time kind of confirmed my hunch of sticking to `UTCTime` and `NominalDiffTime` when possible.

ghost commented:
Thanks for the reference, it makes a bit more sense now! I have personally always used `UTCTime` and `NominalDiffTime` and I know understand why :)
